semgrep-core/src/core/ast/Ugly_print_AST.ml

```diff
@@ -107,12 +137,49 @@ class base_printer : printer_t =
       (* TODO Handle other kinds *)
       | _ -> print_fail ()
 
-    method print_expr { e; _ } = self#print_expr_kind e
+    method print_expr e =
+      let/ res = self#print_expr_without_parens e in
+      (* An alternative would be to simply have the language-specific printers
+       * add parentheses when they implement `print_opcall` and any other
+       * expressions that need parentheses. Doing it here instead has the
+       * advantage that subclasses need only concern themselves with correctly
+       * printing a given node in isolation, and don't need to worry about
+       * whether it needs to be wrapped in parentheses.
+       *
+       * Otherwise, this logic would need to be duplicated in all of the
+       * language printers as well as the hybrid printer used for autofix. There
+       * will surely be some language-specific differences in parenthesis
+       * insertion, but there is enough commonality between languages that much
+       * of the logic should be reusable. *)
+      Ok (self#add_parens_if_needed (G.E e) res)
+
+    method private print_expr_without_parens { e; _ } = self#print_expr_kind e
     method print_expr_kind _ = print_fail ()
     method print_argument _ = print_fail ()
     method print_arguments _ = print_fail ()
 
+    method print_call e args =
+      match e.G.e with
+      (* Binary operator expressions are desugared to calls, but are different
+       * enough that they merit a separate method here. *)
+      | G.IdSpecial (G.Op _, _) -> self#print_opcall e args
```

Comment:
Why not pass the op as an argument instead of e, that way the signature of print_opcall is also clearer.

Reply:
I did this so that print-avoidance would work for operators. Because we can look up expressions and use their original text, when `print_opcall` calls `print_expr`, this just works. If we passed down the operator, it wouldn't. There is no `any` variant for `operator` (and even if there were, it doesn't have location information so it would require some additional tweaking), or is there one for `special` or `special wrap`. Because the external printer works by storing a table with `AST_generic.any` as the key, making this work would require some changes.

Either we can pass down the expression here, which I don't think makes printing any more onerous since you would just call `print_expr` instead of `print_operator`, or add a variant to `AST_generic.any`, or change how the external printer works to allow it to print non-any nodes, which would mean more boilerplate. I prefer this solution, though I think the last may also be worth exploring. I do not think it is a good idea to add an `any` variant to support this.

I'll add a comment explaining this.

Reply:
I'll also add a test that verifies that operators are lifted from the original text (since we can't currently print them, the fact that the existing tests pass are proof of that, but once we implement that, nothing will prevent someone from inadvertently breaking it)